extensions/wikia/Email/Controller/FounderController.class.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+namespace Email\Controller;
+
+use Email\Check;
+use Email\EmailController;
+use Email\Fatal;
+
+abstract class FounderController extends EmailController {
+
+	/** @var \Title */
+	protected $pageTitle;
+
+	protected $previousRevId;
+	protected $currentRevId;
+
+	public function initEmail() {
+		// This title is for the article being commented upon
+		$titleText = $this->request->getVal( 'pageTitle' );
+		$titleNamespace = $this->request->getVal( 'pageNs' );
+
+		$this->pageTitle = \Title::newFromText( $titleText, $titleNamespace );
+
+		$this->previousRevId = $this->request->getVal( 'previousRevId' );
+		$this->currentRevId = $this->request->getVal( 'currentRevId' );
+
+		$this->assertValidParams();
+	}
+
+	/**
+	 * Validate the params passed in by the client
+	 */
+	private function assertValidParams() {
+		$this->assertValidTitle();
+		$this->assertValidRevisionIds();
+	}
+
+	/**
+	 * @throws \Email\Check
+	 */
+	private function assertValidTitle() {
+		if ( !$this->pageTitle instanceof \Title ) {
+			throw new Check( "Invalid value passed for title" );
+		}
+
+		if ( !$this->pageTitle->exists() ) {
+			throw new Check( "Title doesn't exist." );
+		}
+	}
+
+	/**
+	 * @throws \Email\Check
+	 */
+	private function assertValidRevisionIds() {
+		if ( empty( $this->previousRevId ) ) {
+			throw new Check( "Invalid value passed for previousRevId" );
+		}
+
+		if ( empty( $this->currentRevId ) ) {
+			throw new Check( "Invalid value passed for currentRevId" );
+		}
+	}
+
+	/**
+	 * @template avatarLayout
+	 */
+	public function body() {
+		$this->response->setData( [
+			'salutation' => $this->getSalutation(),
+			'editorProfilePage' => $this->getCurrentProfilePage(),
+			'editorUserName' => $this->getCurrentUserName(),
+			'editorAvatarURL' => $this->getCurrentAvatarURL(),
+			'summary' => $this->getSummary(),
+			'buttonText' => $this->getChangesLabel(),
+			'buttonLink' => $this->getChangesLink(),
+			'contentFooterMessages' => [
+				$this->getFooterEncouragement(),
+				$this->getFooterArticleLink(),
+				$this->getFooterAllChangesLink(),
+			],
+			'details' => $this->getDetails(),
+			'hasContentFooterMessages' => true
```

Comment:
Oh, hmm, I think I see what you mean. Like have that logic in one place so we wouldn't have to set it explicitly in each of the controllers (hardcoded or otherwise)? That could be worth looking into. We're actually duplicating quite a bit with all of the `body` methods we keep redefining with similar looking arrays of data. Consolidating that at some point would be nice.